src/main/scala/worker/Master.scala

```diff
@@ -0,0 +1,261 @@
+package worker
+
+import akka.actor.typed.Behavior
+import akka.actor.typed.scaladsl.Behaviors
+import akka.actor.typed.ActorRef
+import akka.cluster.pubsub.{DistributedPubSub, DistributedPubSubMediator}
+import akka.persistence.typed.PersistenceId
+import akka.persistence.typed.scaladsl.Effect
+import akka.persistence.typed.scaladsl.EventSourcedBehavior
+import worker.WorkState.WorkAccepted
+import worker.WorkState.WorkDomainEvent
+import worker.WorkState.WorkerTimedOut
+import akka.actor.typed.scaladsl.adapter._
+import worker.WorkState.WorkCompleted
+import worker.WorkState.WorkStarted
+import worker.WorkState.WorkerFailed
+import worker.Worker.WorkerCommand
+
+import scala.collection.mutable.ListBuffer
+import scala.concurrent.duration.{Deadline, FiniteDuration, _}
+
+/**
+  * The master actor keep tracks of all available workers, and all scheduled and ongoing work items
+  */
+object Master {
+  sealed trait MasterCommand extends CborSerializable
+  // Messages from Workers
+  case class RegisterWorker(workerId: String, replyTo: ActorRef[WorkerCommand])
+      extends MasterCommand
+  case class DeRegisterWorker(workerId: String) extends MasterCommand
+  case class WorkerRequestsWork(workerId: String,
+                                replyTo: ActorRef[Worker.SubmitWork])
+      extends MasterCommand
+  case class WorkIsDone(workerId: String,
+                        workId: String,
+                        result: Any,
+                        replyTo: ActorRef[Worker.Ack])
+      extends MasterCommand
+  case class WorkFailed(workerId: String, workId: String) extends MasterCommand
+
+  private case object Cleanup extends MasterCommand
+
+  // External commands
+  case class SubmitWork(work: Work, replyTo: ActorRef[Master.Ack])
+      extends MasterCommand
+
+  def apply(workTimeout: FiniteDuration): Behavior[MasterCommand] =
+    Behaviors.setup { ctx =>
+      Behaviors.withTimers { timers =>
+        timers.startSingleTimer("cleanup", Cleanup, workTimeout / 2)
+
+        // No typed pub sub yet
+        val mediator = DistributedPubSub(ctx.system.toClassic).mediator
+
+        // the set of available workers is not event sourced as it depends on the current set of workers
+        var workers = Map[String, WorkerState]()
```

Comment:
would it be possible to split this into two actors? one managing the workers and another for the event sourced WorkState? 

They could still be tightly coupled by sharing same messages.

Reply:
Yes I think it is worth trying a different pattern here, the var is easy to get wrong

Reply:
It does get very messy unless we were to change it entirely. As we often update the persistent state and the transient state at the same time

Reply:
Ok, leave it as is then